src/stores/spaces/SpaceStore.ts

```diff
@@ -349,176 +350,335 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
+            this.switchSpaceIfNeeded();
+        }
     };
 
-    private onRoomsUpdate = throttle(() => {
-        // TODO resolve some updates as deltas
+    private rebuildMetaSpaces = () => {
+        const enabledMetaSpaces = new Set(this.enabledMetaSpaces);
         const visibleRooms = this.matrixClient.getVisibleRooms();
 
-        const oldFilteredRooms = this.spaceFilteredRooms;
-        this.spaceFilteredRooms = new Map();
-
-        const enabledMetaSpaces = new Set(this.enabledMetaSpaces);
-        // populate the Home metaspace if it is enabled and is not set to all rooms
-        if (enabledMetaSpaces.has(MetaSpace.Home) && !this.allRoomsInHome) {
-            // put all room invites in the Home Space
-            const invites = visibleRooms.filter(r => !r.isSpaceRoom() && r.getMyMembership() === "invite");
-            this.spaceFilteredRooms.set(MetaSpace.Home, new Set(invites.map(r => r.roomId)));
-
-            visibleRooms.forEach(room => {
-                if (this.showInHomeSpace(room)) {
-                    this.spaceFilteredRooms.get(MetaSpace.Home).add(room.roomId);
-                }
-            });
+        if (enabledMetaSpaces.has(MetaSpace.Home)) {
+            this.rebuildHomeSpace();
+        } else {
+            this.spaceFilteredRooms.delete(MetaSpace.Home);
         }
 
-        // populate the Favourites metaspace if it is enabled
         if (enabledMetaSpaces.has(MetaSpace.Favourites)) {
             const favourites = visibleRooms.filter(r => r.tags[DefaultTagID.Favourite]);
             this.spaceFilteredRooms.set(MetaSpace.Favourites, new Set(favourites.map(r => r.roomId)));
+        } else {
+            this.spaceFilteredRooms.delete(MetaSpace.Favourites);
         }
 
-        // populate the People metaspace if it is enabled
-        if (enabledMetaSpaces.has(MetaSpace.People)) {
-            const people = visibleRooms.filter(r => DMRoomMap.shared().getUserIdForRoomId(r.roomId));
-            this.spaceFilteredRooms.set(MetaSpace.People, new Set(people.map(r => r.roomId)));
-        }
+        // The People metaspace doesn't need maintaining
 
-        // populate the Orphans metaspace if it is enabled
-        if (enabledMetaSpaces.has(MetaSpace.Orphans)) {
+        // Populate the orphans space if the Home space is enabled as it is a superset of it.
+        // Home is effectively a super set of People + Orphans with the addition of having all invites too.
+        if (enabledMetaSpaces.has(MetaSpace.Orphans) || enabledMetaSpaces.has(MetaSpace.Home)) {
             const orphans = visibleRooms.filter(r => {
                 // filter out DMs and rooms with >0 parents
                 return !this.parentMap.get(r.roomId)?.size && !DMRoomMap.shared().getUserIdForRoomId(r.roomId);
             });
             this.spaceFilteredRooms.set(MetaSpace.Orphans, new Set(orphans.map(r => r.roomId)));
         }
 
+        if (this.activeSpace[0] !== "!") {
```

Comment:
This needs a comment - what does `!` denote?

Reply:
actually, is this getting the first character of the potential room ID? Can we not do that please?

Reply:
its either a room ID if a real space is selected or an enum member if a meta space is selected

Reply:
Really not comfortable with this sort of check for whether or not something references a room :( 

`!matrixClient.getRoom(this.activeSpace)` would be better, but also not great. It should be just as fast for the purposes of the functions here.

Reply:
will make a utility function which explicitly checks against all enum members instead, as a room ID of a room your client doesn't know about might be valid one day